include/boost/beast/core/impl/buffers_adaptor.hpp

```diff
@@ -34,332 +35,162 @@ namespace beast {
-        ++(*this);
-        return temp;
-    }
+private:
+    void
+    adjust(
+        std::size_t pos,
+        std::size_t n);
 
-    const_iterator&
-    operator--()
-    {
-        --it_;
-        return *this;
-    }
+private:
+    // points to the first buffer in the sequence
+    iter_type first_;
 
-    const_iterator
-    operator--(int)
-    {
-        auto temp = *this;
-        --(*this);
-        return temp;
-    }
+    // Points to one past the end of the underlying buffer sequence
+    iter_type last_;
 
-private:
-    friend class readable_bytes;
+    // The initial offset into the first buffer
+    std::size_t first_offset_;
 
-    const_iterator(
-        buffers_adaptor const& b,
-        iter_type iter)
-        : it_(iter)
-        , b_(&b)
-    {
-    }
+    // how many bytes in the penultimate buffer are used (if any)
+    std::size_t last_size_;
 };
 
-template<class MutableBufferSequence>
-template<bool isMutable>
-auto
-buffers_adaptor<MutableBufferSequence>::
-readable_bytes<isMutable>::
-begin() const ->
-    const_iterator
-{
-    return const_iterator{*b_, b_->begin_};
-}
-
-template<class MutableBufferSequence>
-template<bool isMutable>
-auto
-buffers_adaptor<MutableBufferSequence>::
-readable_bytes<isMutable>::
-readable_bytes::end() const ->
-    const_iterator
-{
-    return const_iterator{*b_, b_->end_impl()};
-}
+#if BOOST_WORKAROUND(BOOST_MSVC, < 1910)
+# pragma warning (pop)
+#endif
 
 //------------------------------------------------------------------------------
 
 template<class MutableBufferSequence>
-class buffers_adaptor<MutableBufferSequence>::
-mutable_buffers_type
+template<bool isMutable>
+struct buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+iterator
 {
-    buffers_adaptor const* b_;
-
-public:
-    using value_type = net::mutable_buffer;
-
-    class const_iterator;
-
-    mutable_buffers_type() = delete;
-    mutable_buffers_type(
-        mutable_buffers_type const&) = default;
-    mutable_buffers_type& operator=(
-        mutable_buffers_type const&) = default;
-
-    const_iterator
-    begin() const;
-
-    const_iterator
-    end() const;
-
-private:
-    friend class buffers_adaptor;
-
-    mutable_buffers_type(
-            buffers_adaptor const& b)
-        : b_(&b)
+    struct iterator_category
+        : std::bidirectional_iterator_tag
     {
-    }
-};
-
-template<class MutableBufferSequence>
-class buffers_adaptor<MutableBufferSequence>::
-mutable_buffers_type::const_iterator
-{
-    iter_type it_{};
-    buffers_adaptor const* b_ = nullptr;
-
-public:
-    using value_type = net::mutable_buffer;
-    using pointer = value_type const*;
-    using reference = value_type;
+    };
+    using value_type = typename buffers_adaptor<MutableBufferSequence>::
+    template subrange<isMutable>::
+    value_type;
+    using reference = typename std::add_lvalue_reference<value_type>::type;
+    using pointer = typename std::add_pointer<value_type>::type;
```

Comment:
There's no need for using `add_pointer` or `add_lvalue_reference` why did you do that?